File: apps/web/src/lib/call.ts
/**
 * The media layer, behind an interface narrow enough to be replaced by nothing.
 *
 * # Why this file exists at all
 *
 * `session-call.ts` holds the whole of what a call *means* — who is ringing, who answered, when
 * it started, when it may be given up on. None of that involves a media server, and all of it is
 * worth testing. Importing the SDK there would make it untestable by `node --test`, which is the
 * same reason `cipher.ts` hands `Session` a capability instead of a key: a module that cannot be
 * loaded outside a browser is a module nobody writes a test for.
 *
 * So this is a port, and {@link liveMedia} is the only implementation that talks to a network.
 * The state machine takes a {@link Media} and never learns what is behind it.
 *
 * # What the media server is trusted with, and what it is not
 *
 * It routes the audio, and it cannot hear it. Every frame is encrypted a second time, inside the
 * browser, under a key derived from the current MLS epoch — `Client.callKey`, which every member
 * computes locally and nobody sends anywhere. What reaches the server is already ciphertext when
 * the transport encryption is applied to it, and it is still ciphertext when the transport
 * encryption is removed on the far side.
 *
 * What it does learn is real and is not hidden here: who shares a room with whom, and for how
 * long. `crates/server/src/call.rs` argues the whole trade, and `docs/THREAT-MODEL.md` records
 * it where a user might look.
 *
 * # Why the identity is derived rather than given
 *
 * A participant needs a name inside the room, and the two obvious candidates both leak: a device
 * id hands the media server the directory this project spends its effort keeping from the
 * delivery service, and a random name leaves members unable to tell each other apart. So it is
 * derived from the call key — every member can compute every other member's, and the server sees
 * a string that changes at every call and means nothing to it.
 *
 * That derivation is not authentication, and must not be read as one: a member of the group can
 * compute another member's identity and take it. It is the same forgery the ephemeral channel
 * allows, from the same cause — a key that belongs to the group authenticates the group.
 */
import type { Room } from "livekit-client";

/**
 * Whether this build was told where a media server is.
 *
 * A *build-time* value, like the policy in `csp.ts` that has to name the same origin — a client
 * whose Content-Security-Policy does not carry that origin cannot reach it, so a runtime toggle
 * would offer a button that is guaranteed to fail. The server has the last word regardless: it
 * answers 503 when it holds no credentials, and the call ends there.
 *
 * Read defensively because this module is also loaded by `node --test`, where `import.meta.env`
 * does not exist.
 */
export const CALLS_CONFIGURED = Boolean(
  (import.meta as { env?: Record<string, string | undefined> }).env?.VITE_MEDIA_URL,
);

/** Where a call happens, and what it takes to be let in. */
export interface Admission {
  /** The media server, as the delivery service knows it. */
  url: string;
  token: string;
  /** The relay, when the deployment has one. */
  relay?: { urls: string[]; username: string; credential: string };
}

/** What a joined call offers. Deliberately four verbs: anything more is state, and state is not here. */
export interface Call {
  /** Silences or restores the microphone. */
  mute(muted: boolean): Promise<void>;
  /**
   * Hands the call a new key.
   *
   * Called on every MLS commit. Without it a call spanning an epoch change goes silent — which
   * is the correct failure for a member who has just been removed, and the wrong one for the
   * five who have not.
   */
  rekey(key: Uint8Array): Promise<void>;
  /** Leaves. Idempotent: hanging up twice is a thing interfaces do. */
  leave(): Promise<void>;
  /** The identities currently publishing audio, ours excluded. */
  peers(): string[];
}

/** How a call is joined. The state machine holds one of these and nothing else about media. */
export interface Media {
  join(options: JoinOptions): Promise<Call>;
}

export interface JoinOptions {
  admission: Admission;
  /** The frame key, derived from the MLS epoch. Never leaves this device. */
  key: Uint8Array;
  /** Called whenever the participant list changes, with our own identity excluded. */
  onPeers: (peers: string[]) => void;
  /**
   * Called when the connection is lost for good.
   *
   * A call that dies has to reach the state machine, or the interface keeps showing a
   * conversation that ended — which is worse than showing none.
   */
  onClosed: () => void;
}

/**
 * The name a device answers to inside a room.
 *
 * Derived from the call key so that members recognise each other while the media server sees an
 * opaque string. Truncated to sixteen hex characters: it names one participant of one call, and
 * a full digest would only be longer.
 */
export async function identityFor(key: Uint8Array, deviceId: string): Promise<string> {
  const material = new TextEncoder().encode(`wac-call-peer-v1${deviceId}`);
  const input = new Uint8Array(key.length + material.length);
  input.set(key, 0);
  input.set(material, key.length);

  const digest = new Uint8Array(await crypto.subtle.digest("SHA-256", input.slice().buffer));
  return [...digest.subarray(0, 8)].map((byte) => byte.toString(16).padStart(2, "0")).join("");
}

/**
 * The real media layer.
 *
 * Imported lazily, and that is not a performance flourish: the SDK and its encryption worker are
 * a few hundred kilobytes that a user who never places a call has no reason to download, and
 * this application is served to people whose connection is the reason they chose it.
 */
export function liveMedia(): Media {
  return {
    async join(options: JoinOptions): Promise<Call> {
      const { ExternalE2EEKeyProvider, Room, RoomEvent, Track } = await import("livekit-client");

      const keys = new ExternalE2EEKeyProvider();
      await keys.setKey(options.key.slice().buffer);

      const room = new Room({
        adaptiveStream: false,
        dynacast: false,
        e2ee: {
          keyProvider: keys,
          // Same-origin, bundled by Vite. The SDK otherwise reaches for a `blob:` wrapper, which
          // `worker-src 'self'` refuses — silently, as a blocked worker reports no reason.
          worker: new Worker(new URL("livekit-client/e2ee-worker", import.meta.url), {
            type: "module",
          }),
        },
        ...(options.admission.relay && {
          rtcConfig: {
            iceServers: [
              {
                urls: options.admission.relay.urls,
                username: options.admission.relay.username,
                credential: options.admission.relay.credential,
              },
            ],
          },
        }),
      });

      const announce = () => options.onPeers(peersOf(room));

      // Subscribed **before** connecting, all of them. A listener registered afterwards misses
      // whoever was already in the room — which is every participant, for the second person to
      // join a call.
      room
        .on(RoomEvent.ParticipantConnected, announce)
        .on(RoomEvent.ParticipantDisconnected, announce)
        .on(RoomEvent.Disconnected, options.onClosed)
        // Audio has to be attached to an element to be heard. Nothing is rendered: an `<audio>`
        // element with no controls is the browser's own way of playing a stream, and giving React
        // a node to own would tie the sound to a component's lifetime rather than the call's.
        //
        // **Every track is attached, including one from an account the user has blocked, and
        // that is decided rather than missed.**
        //
        // Blocking closes the channels that reach somebody without being asked for: the thread,
        // the typing indicator, notifications, and the ring. A group call is not one of them.
        // Joining is a deliberate act — the room is entered on purpose, by somebody who can see
        // who is in it — so what happens inside it is not content arriving unbidden, and blocking
        // has no claim over it.
        //
        // The line is there, and it is worth naming because it is the one to reason from: **what
        // arrives without being asked for is filtered; what is walked into is not.**
        //
        // So the test is on **who placed the call**, never on who is in it. An invitation from a
        // blocked account is refused in `session.ts` and never rings — that one arrives unbidden,
        // and it is the whole of the exception. A blocked member sitting in a call somebody else
        // placed is not: this device chose to enter that room. Same if they join after us, or if
        // the person who called leaves and they stay; the question was settled when the ring was
        // answered, and re-asking it mid-call would end a conversation under its participants for
        // a reason none of them could see.
        //
        // Filtering here would also cost the wrong person. Muting one voice in a conversation
        // several people are holding breaks it for the one who muted — the others answer somebody
        // this device cannot hear, and the thread of what is being said is lost. The blocked
        // member loses nothing. A measure whose whole effect lands on the person who took it is
        // not protection, it is a penalty for having used the feature.
        //
        // Nothing is given up on confidentiality by this: that member holds the epoch key
        // legitimately, being in the group at all is what the call is derived from, and removing
        // them from the group is the action that actually ends their access — see
        // `Session.tickCall`.
        //
        // **One debt this decision creates, and it is owed to the reader of another screen.** The
        // blocking control says "hides what someone says", which stopped being true the moment
        // calls existed: a blocked member's voice arrives in a call somebody else placed. That
        // wording lives outside this branch, so it cannot be fixed here — it has to change to
        // "hides what someone writes", plus a line saying their voice still carries in a call
        // they did not place, at the same time as the ring guard lands in `session.ts`. A control
        // describing an effect it does not have is the defect this whole comment exists to avoid
        // creating.
        .on(RoomEvent.TrackSubscribed, (track) => {
          if (track.kind === Track.Kind.Audio) track.attach();
          announce();
        });

      await room.connect(options.admission.url, options.admission.token);

      // **This line is what makes the call end-to-end encrypted, and it is not implied by the
      // `e2ee` option above.**
      //
      // Passing `e2ee` to the constructor builds the key provider, loads the worker and accepts
      // every key handed to it. It does not encrypt anything: the SDK only sets the local
      // participant's encryption type when this is called, so without it the frames reach the
      // media server in the clear and the derived key does nothing at all. Nothing reports the
      // difference — the call connects, the audio is heard, and every key rotation appears to
      // work, because none of it is in the media path.
      //
      // That is exactly how it was found: a member removed from the group mid-call went on
      // hearing it for forty seconds. See `docs/THREAT-MODEL.md` for what this line is
      // load-bearing for; the short of it is everything this module claims.
      //
      // **After `connect` and before publishing.** Before connecting, the local participant has
      // no identity yet and the SDK skips enabling its cryptor; after publishing, turning it on
      // republishes every track, which drops the audio it has just started carrying.
      await room.setE2EEEnabled(true);

      await room.localParticipant.setMicrophoneEnabled(true);

      announce();

      return {
        async mute(muted: boolean) {
          await room.localParticipant.setMicrophoneEnabled(!muted);
        },
        async rekey(key: Uint8Array) {
          await keys.setKey(key.slice().buffer);
        },
        async leave() {
          await room.disconnect();
        },
        peers: () => peersOf(room),
      };
    },
  };
}

function peersOf(room: Room): string[] {
  return [...room.remoteParticipants.values()].map((participant) => participant.identity);
}

File: crates/server/src/call.rs
//! Admission to a call: the access token, and where to reach the relay.
//!
//! # What this module gives up, before anything else
//!
//! The rest of this server routes opaque blobs and can decrypt none of them. Audio cannot work
//! that way. A media server has to read the transport in order to route one stream to five
//! listeners without holding five conversations, so it terminates the transport encryption —
//! and a participant's stream carries a stable identity for the length of a call, where a posted
//! envelope carries none at all.
//!
//! **So a call leaks more than a message does**, and the two places it leaks are here:
//!
//! * this server signs the token, so it sees that somebody is joining a call, when, and towards
//!   which group;
//! * the media server sees who is connected to the same room as whom, and for how long.
//!
//! What neither of them sees is the audio. It is encrypted frame by frame under a key derived
//! from the MLS epoch — `Conversation::call_key`, in `crypto-core` — which is never sent
//! anywhere: every member computes it locally. There is no point at which this server or the
//! media server could be asked for it, because neither is ever in possession of one.
//!
//! `docs/THREAT-MODEL.md` carries the same statement where a reader will find it. This one is
//! here for whoever changes this file.
//!
//! # Two things are deliberately withheld from the media server
//!
//! **The group id.** The room is named by a digest over the group id and the call id, so the
//! relay learns that some room exists, never which conversation it belongs to. Two calls in the
//! same conversation are two unrelated rooms as far as it can tell.
//!
//! **The device id.** The identity in the token is chosen by the caller and this server does not
//! check it — it cannot, the route being authenticated by the group MAC rather than by a
//! signature, which is what keeps a call from revealing who placed it. Clients derive that
//! identity from the call key, so members recognise each other and the relay sees an opaque
//! string that changes every call. A member could take another member's identity inside a room,
//! which is the same forgery the ephemeral channel already allows and documents: a group key
//! authenticates the group, not the member.
//!
//! # Inert without configuration
//!
//! [`Media::from_environment`] returns a configuration with nothing in it when the variables are
//! absent, and the route then answers 503. A deployment that runs no media server keeps a fully
//! working messenger, exactly as one that talks to neither Apple nor Google keeps one — see
//! [`crate::push`], which this follows deliberately.

use std::time::{SystemTime, UNIX_EPOCH};

use base64::Engine as _;
use base64::engine::general_purpose::{STANDARD as BASE64_STANDARD, URL_SAFE_NO_PAD};
use hmac::{Hmac, Mac};
use sha2::{Digest, Sha256};

/// How long an access token stays valid.
///
/// It is admission to a room, not the length of a call: once joined, the media server keeps the
/// connection. Long enough that a slow client still gets in, short enough that a token captured
/// from a log is worthless by the time anyone reads it.
const TOKEN_TTL_SECONDS: u64 = 300;

/// How long a relay credential stays valid.
///
/// Longer than the token, because it is needed again at every network change — a phone moving
/// from Wi-Fi to cellular re-runs candidate gathering, and a credential that expired mid-call
/// would drop the audio precisely when the connection was already struggling.
const RELAY_TTL_SECONDS: u64 = 12 * 3600;

/// The identity a client may ask for, in bytes.
///
/// The server cannot verify it — see the module header — so all it can do is refuse one large
/// enough to be used as storage.
const MAX_IDENTITY_BYTES: usize = 64;

/// The call id, in bytes. Clients send a hex digest; this only stops it being used as a field.
const MAX_CALL_ID_BYTES: usize = 64;

/// Where the media lives, if anywhere.
///
/// Both halves are optional and independent: a deployment may run a media server reachable
/// without a relay, and one behind a relay it does not own.
#[derive(Clone, Default)]
pub struct Media {
    pub sfu: Option<Sfu>,
    pub relay: Option<Relay>,
}

/// The media server, and the credentials to mint tokens for it.
#[derive(Clone)]
pub struct Sfu {
    /// The address clients connect to. Handed out as-is: this server never contacts it.
    pub url: String,
    pub api_key: String,
    pub api_secret: String,
}

/// The relay that carries the media when a direct path cannot be found.
#[derive(Clone)]
pub struct Relay {
    pub urls: Vec<String>,
    /// Shared with the relay, and with nothing else. It mints credentials, it is never sent.
    pub secret: String,
}

impl Media {
    /// Reads the configuration, and is content to find none.
    ///
    /// A partial configuration is treated as no configuration for that half, and says so in the
    /// log. The alternative — starting with a key and no secret — fails at the first call, on a
    /// path nobody is watching.
    pub fn from_environment() -> Self {
        let read = |name: &str| std::env::var(name).ok().filter(|value| !value.is_empty());

        let sfu = match (read("MEDIA_URL"), read("MEDIA_API_KEY"), read("MEDIA_API_SECRET")) {
            (Some(url), Some(api_key), Some(api_secret)) => Some(Sfu { url, api_key, api_secret }),
            (None, None, None) => None,
            _ => {
                tracing::warn!("media server half-configured: calls stay off");
                None
            }
        };

        let relay = match (read("RELAY_URLS"), read("RELAY_SECRET")) {
            (Some(urls), Some(secret)) => Some(Relay {
                urls: urls.split(',').map(|url| url.trim().to_owned()).collect(),
                secret,
            }),
            (None, None) => None,
            _ => {
                tracing::warn!("relay half-configured: calls will only work on a direct path");
                None
            }
        };

        Self { sfu, relay }
    }
}

/// The room a call happens in.
///
/// A digest rather than the group id: the media server has no use for the identifier this whole
/// server is organised around, and handing it over would let it group a deployment's calls into
/// conversations. The call id is in the digest too, so two calls in one conversation do not look
/// related either.
pub fn room_name(group_id: &[u8], call_id: &str) -> String {
    let mut digest = Sha256::new();
    digest.update(b"wac-call-room-v1");
    digest.update(group_id);
    digest.update(call_id.as_bytes());
    hex::encode(digest.finalize())
}

impl Sfu {
    /// Mints an access token for one room.
    ///
    /// The permissions are the narrowest that let a call happen: join, publish, subscribe. Not
    /// data — the conversation already has an encrypted channel of its own, and a second one
    /// through the relay would be a second one to secure.
    pub fn token(&self, room: &str, identity: &str) -> String {
        let now = seconds_now();

        let header = serde_json::json!({ "alg": "HS256", "typ": "JWT" });
        let claims = serde_json::json!({
            "iss": self.api_key,
            "sub": identity,
            "nbf": now,
            "exp": now + TOKEN_TTL_SECONDS,
            "video": {
                "room": room,
                "roomJoin": true,
                "canPublish": true,
                "canSubscribe": true,
                "canPublishData": false,
            },
        });

        let signing = format!("{}.{}", encode_part(&header), encode_part(&claims));
        let mut mac = Hmac::<Sha256>::new_from_slice(self.api_secret.as_bytes())
            .expect("HMAC accepts a key of any length");
        mac.update(signing.as_bytes());

        format!("{signing}.{}", URL_SAFE_NO_PAD.encode(mac.finalize().into_bytes()))
    }
}

impl Relay {
    /// Mints a short-lived relay credential.
    ///
    /// The scheme is the relay's own: the user name is the expiry, and the password is a MAC
    /// over it under the shared secret. It means no account exists anywhere — nothing to
    /// provision, nothing to revoke, and a captured credential dies on its own.
    ///
    /// **The MAC is SHA-1 because the relay only speaks SHA-1 here.** It authenticates a name
    /// that expires in hours and protects nothing at rest; the alternative is no credential at
    /// all, since the other end would reject anything else.
    pub fn credential(&self) -> (String, String) {
        let username = (seconds_now() + RELAY_TTL_SECONDS).to_string();

        let mut mac = Hmac::<sha1::Sha1>::new_from_slice(self.secret.as_bytes())
            .expect("HMAC accepts a key of any length");
        mac.update(username.as_bytes());

        (username, BASE64_STANDARD.encode(mac.finalize().into_bytes()))
    }
}

/// Refuses an identity or a call id that could be used as a field rather than as a name.
pub fn acceptable(call_id: &str, identity: &str) -> bool {
    !call_id.is_empty()
        && call_id.len() <= MAX_CALL_ID_BYTES
        && !identity.is_empty()
        && identity.len() <= MAX_IDENTITY_BYTES
        && call_id.bytes().all(|byte| byte.is_ascii_alphanumeric())
        && identity.bytes().all(|byte| byte.is_ascii_alphanumeric())
}

fn encode_part(value: &serde_json::Value) -> String {
    URL_SAFE_NO_PAD.encode(serde_json::to_vec(value).expect("a JSON object always serialises"))
}

fn seconds_now() -> u64 {
    SystemTime::now().duration_since(UNIX_EPOCH).map(|since| since.as_secs()).unwrap_or(0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_room_hides_the_group_it_belongs_to() {
        let room = room_name(b"group-1", "abc");

        assert_eq!(room.len(), 64, "a SHA-256 digest, hex");
        assert!(!room.contains(&hex::encode(b"group-1")), "the group id must not survive");
    }

    /// Two calls in one conversation must not look related to the relay, or the digest would
    /// only be hiding the identifier while keeping the grouping it enables.
    #[test]
    fn two_calls_in_one_group_are_two_unrelated_rooms() {
        assert_ne!(room_name(b"group-1", "abc"), room_name(b"group-1", "def"));
    }

    #[test]
    fn a_token_carries_the_room_and_expires() {
        let sfu = Sfu {
            url: "wss://media.example".into(),
            api_key: "key".into(),
            api_secret: "secret".into(),
        };

        let token = sfu.token("room-1", "peer-1");
        let parts: Vec<&str> = token.split('.').collect();
        assert_eq!(parts.len(), 3, "header, claims, signature");

        let claims: serde_json::Value =
            serde_json::from_slice(&URL_SAFE_NO_PAD.decode(parts[1]).unwrap()).unwrap();

        assert_eq!(claims["video"]["room"], "room-1");
        assert_eq!(claims["sub"], "peer-1");
        assert_eq!(claims["iss"], "key");
        assert_eq!(
            claims["exp"].as_u64().unwrap() - claims["nbf"].as_u64().unwrap(),
            TOKEN_TTL_SECONDS
        );
    }

    /// The data channel is refused on purpose: the conversation already has an encrypted channel,
    /// and a second one through the relay would be a second one to secure.
    #[test]
    fn a_token_grants_no_data_channel() {
        let sfu = Sfu { url: "u".into(), api_key: "k".into(), api_secret: "s".into() };
        let token = sfu.token("room-1", "peer-1");
        let claims: serde_json::Value = serde_json::from_slice(
            &URL_SAFE_NO_PAD.decode(token.split('.').nth(1).unwrap()).unwrap(),
        )
        .unwrap();

        assert_eq!(claims["video"]["canPublishData"], false);
    }

    /// A different secret has to produce a different signature, or the token would be a claim
    /// anybody could make.
    #[test]
    fn the_signature_depends_on_the_secret() {
        let one = Sfu { url: "u".into(), api_key: "k".into(), api_secret: "one".into() };
        let two = Sfu { url: "u".into(), api_key: "k".into(), api_secret: "two".into() };

        let signature = |token: String| token.split('.').nth(2).unwrap().to_owned();

        assert_ne!(signature(one.token("room", "peer")), signature(two.token("room", "peer")));
    }

    #[test]
    fn a_relay_credential_is_a_mac_over_its_own_expiry() {
        let relay = Relay { urls: vec!["turn:relay.example".into()], secret: "shared".into() };
        let (username, password) = relay.credential();

        let expiry: u64 = username.parse().expect("the user name is the expiry");
        assert!(expiry > seconds_now(), "an already expired credential opens nothing");

        let mut mac = Hmac::<sha1::Sha1>::new_from_slice(b"shared").unwrap();
        mac.update(username.as_bytes());
        assert_eq!(password, BASE64_STANDARD.encode(mac.finalize().into_bytes()));
    }

    #[test]
    fn an_identifier_that_is_not_a_name_is_refused() {
        assert!(acceptable("9f2c41ab", "peer1"));
        assert!(!acceptable("", "peer1"));
        assert!(!acceptable("9f2c41ab", ""));
        assert!(!acceptable("../etc", "peer1"), "a path is not a call id");
        assert!(!acceptable("9f2c41ab", &"a".repeat(MAX_IDENTITY_BYTES + 1)));
    }
}
